controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/metal3-io/baremetal-operator/pkg/utils"
+)
+
+const (
+	subscriptionRetryDelay = time.Minute * 10
+)
+
+type BMCEventSubscriptionReconciler struct {
+	client.Client
+	Log                logr.Logger
+	ProvisionerFactory provisioner.Factory
+	APIReader          client.Reader
+}
+
+//+kubebuilder:rbac:groups=metal3.io,resources=bmceventsubscriptions,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=bmceventsubscriptions/status,verbs=get;update;patch
+
+func (r *BMCEventSubscriptionReconciler) Reconcile(ctx context.Context, request ctrl.Request) (result ctrl.Result, err error) {
+	reqLogger := r.Log.WithValues("bmceventsubscription", request.NamespacedName)
+	reqLogger.Info("start")
+
+	defer func() {
+		reqLogger.Info("done")
+	}()
+
+	// Fetch the BMCEventSubscription
+	subscription := &metal3v1alpha1.BMCEventSubscription{}
+	err = r.Get(ctx, request.NamespacedName, subscription)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after
+			// reconcile request.  Owned objects are automatically
+			// garbage collected. For additional cleanup logic use
+			// finalizers.  Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load subscription")
+	}
+
+	host := &metal3v1alpha1.BareMetalHost{}
+	namespacedHostName := types.NamespacedName{
+		Name:      subscription.Spec.HostName,
+		Namespace: request.Namespace,
+	}
+	err = r.Get(ctx, namespacedHostName, host)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			reqLogger.Error(err, "baremetalhost not found", "host", subscription.Spec.HostName)
+
+			subscription.Status.Error = fmt.Sprintf("baremetal host \"%s\"", subscription.Status.Error)
+			err := r.Status().Update(ctx, subscription)
+
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription status")
+			}
+
+			return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load host data")
+	}
+
+	err = r.addFinalizer(ctx, subscription)
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed add finalizer")
+	}
+
+	prov, ready, err := r.getProvisioner(request, host)
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+
+	if !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	if subscription.DeletionTimestamp.IsZero() {
+		// Not being deleted
+		if err := r.createSubscription(ctx, prov, subscription); err != nil {
+			subscription.Status.Error = fmt.Sprintf("failed to create a subscription")
+			err := r.Status().Update(ctx, subscription)
```

Comment:
This `err` will cover the previous one at line 120, so in case the create subscription will fail (but the update not) the reconcile loop will return without any error